ClassFieldTheory/Mathlib/RingTheory/HenselPolynomial.lean

```diff
@@ -0,0 +1,783 @@
+
+open Ideal Quotient
+
+theorem Ideal.Quotient.ker_factorPow_pow {R : Type*} [CommRing R] (I : Ideal R) {m n : ℕ}
+    (le : m ≤ n) (hm : m ≠ 0) : RingHom.ker (factorPow I le) ^ n = ⊥ := by
+  rw [factorPow, factor_ker, ← Ideal.map_pow, ← pow_mul,
+    Ideal.map_mk_eq_bot_of_le (Ideal.pow_le_pow_right <| Nat.le_mul_of_pos_left _ <| by grind)]
+
+theorem IsNilpotent.ker_factorPow {R : Type*} [CommRing R] (I : Ideal R) {m n : ℕ} (le : m ≤ n)
+    (hm : m ≠ 0) : IsNilpotent (RingHom.ker (factorPow I le)) :=
+  ⟨n, by simpa using ker_factorPow_pow I le hm⟩
+
+theorem IsNilpotent.ker_factor {R : Type*} [CommRing R] (I : Ideal R) {n : ℕ} (hn : n ≠ 0) :
+    IsNilpotent (RingHom.ker (factor (I.pow_le_self hn))) :=
+  ⟨n, by rw [factor_ker, ← Ideal.map_pow, map_quotient_self, zero_eq_bot]⟩
+
+theorem Ideal.map_le_nilradical {R S : Type*} [CommSemiring R] [CommSemiring S] {I : Ideal R}
+    (h : I ≤ nilradical R) {f : R →+* S} : I.map f ≤ nilradical S :=
+  map_le_of_le_comap fun _ hx ↦ IsNilpotent.map (h hx) _
+
+/-- If `p ∈ I[X]` then `p /ₘ q ∈ I[X]`. -/
+theorem Polynomial.divByMonic_mem_map {R : Type*} [CommRing R] {I : Ideal R} {p q : R[X]}
+    (hp : p ∈ I.map C) : p /ₘ q ∈ I.map C := by
+  rw [← mk_ker (I := I), ← ker_mapRingHom, RingHom.mem_ker, coe_mapRingHom] at hp ⊢
+  by_cases hq : q.Monic
+  · rw [map_divByMonic _ hq, hp, zero_divByMonic]
+  · rw [divByMonic_eq_of_not_monic _ hq, Polynomial.map_zero]
+
+/-- If `p ∈ I[X]` then `p %ₘ q ∈ I[X]`. -/
+theorem Polynomial.modByMonic_mem_map {R : Type*} [CommRing R] {I : Ideal R} {p q : R[X]}
+    (hp : p ∈ I.map C) : p %ₘ q ∈ I.map C := by
+  rw [← mk_ker (I := I), ← ker_mapRingHom, RingHom.mem_ker, coe_mapRingHom] at hp ⊢
+  by_cases hq : q.Monic
+  · rw [map_modByMonic _ hq, hp, zero_modByMonic]
+  · rw [modByMonic_eq_of_not_monic _ hq, hp]
+
+theorem Polynomial.degree_add_lt {R : Type*} [Semiring R] {p q : R[X]} {n}
+    (hp : p.degree < n) (hq : q.degree < n) : (p + q).degree < n :=
+  (degree_add_le _ _).trans_lt <| max_lt hp hq
+
+theorem Polynomial.degree_sub_lt' {R : Type*} [Ring R] {p q : R[X]} {n}
+    (hp : p.degree < n) (hq : q.degree < n) : (p - q).degree < n :=
+  degree_add_lt hp <| degree_neg q ▸ hq
+
+theorem Polynomial.Monic.degree_mul' {R : Type*} [Semiring R] {p q : Polynomial R} (hp : p.Monic) :
+    (p * q).degree = p.degree + q.degree := by
+  rw [hp.degree_mul_comm, hp.degree_mul, add_comm]
+
+section
+variable {R S : Type*} [Ring R] [Semiring S] {f : R →+* S} (hf : Function.Surjective ⇑f)
+
+theorem RingHom.quotientKerEquivOfSurjective_comp :
+    (RingHom.quotientKerEquivOfSurjective hf : _ →+* _).comp (Ideal.Quotient.mk (ker f)) = f := rfl
+
+theorem RingHom.quotientKerEquivOfSurjective_symm_comp :
+    ((RingHom.quotientKerEquivOfSurjective hf).symm : _ →+* _).comp f =
+    Ideal.Quotient.mk (ker f) := by
+  conv => enter [1,2]; rw [← quotientKerEquivOfSurjective_comp hf]
+  rw [← comp_assoc, RingEquiv.symm_comp, id_comp]
+
+end
+
+end Lemmas
+
```

Comment:
```suggestion
```

Reply:
I do not understand how suggestions work in this new interface. I just wanted to suggest that you remove the second blank line.